x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyLocks.java

```diff
@@ -0,0 +1,71 @@
+package org.elasticsearch.xpack.enrich;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
+
+class EnrichPolicyLocks {
+
+    private final ReadWriteLock coordinationLock = new ReentrantReadWriteLock(true);
+    private final ConcurrentHashMap<String, Semaphore> policyLocks = new ConcurrentHashMap<>();
+    private final AtomicLong policyRunCounter = new AtomicLong(0L);
+
+    static class LockState {
+        final boolean runningPolicies;
+        final long revision;
+
+        LockState(boolean safe, long revision) {
+            this.runningPolicies = safe;
+            this.revision = revision;
+        }
+    }
+
+    void lockPolicy(String policyName) {
+        coordinationLock.readLock().lock();
+        try {
+            Semaphore runLock = policyLocks.computeIfAbsent(policyName, (name) -> new Semaphore(1));
+            boolean acquired = runLock.tryAcquire();
+            if (acquired == false) {
+                throw new EsRejectedExecutionException("Policy execution failed. Policy execution for [" + policyName +
+                    "] is already in progress.");
+            }
+            policyRunCounter.incrementAndGet();
+        } finally {
+            coordinationLock.readLock().unlock();
+        }
+    }
+
+    LockState lockState() {
+        if (coordinationLock.writeLock().tryLock()) {
+            try {
+                long revision = policyRunCounter.get();
+                int currentPolicyExecutions = policyLocks.size();
+                return new LockState(currentPolicyExecutions > 0, revision);
+            } finally {
+                coordinationLock.writeLock().unlock();
+            }
+        }
+        return new LockState(false, policyRunCounter.get());
+    }
+
+    boolean isSafe(LockState previousState) {
+        if (previousState.runningPolicies == true) {
+            return false;
+        }
+        LockState currentState = lockState();
+        return currentState.runningPolicies == false && currentState.revision == previousState.revision;
+    }
+
+    void releasePolicy(String policyName) {
+        coordinationLock.readLock().lock();
+        try {
+            policyLocks.remove(policyName);
```

Comment:
Since the semaphore is only acquired in a non-waiting fashion and only really meant to ensure a first come first serve model, it's just discarded when the policy is completed to keep the concurrent map tidy.

This also protects against any instruction interleaving funny business that might occur:
* Thread 1: gets semaphore from map, preempted by thread scheduler
* Thread 2: releases semaphore, removes semaphore from map, preempted by thread scheduler
* Thread 1: successfully acquires dead semaphore, starts execution
* Background cleanup thread: starts cleanup, sees no new operations occurring on locks, deletes Thread 1's work in progress index

Even if we were to remove from the map first before releasing the semaphore in step 2, it leads to the same scenario. We could just recheck the map after acquiring the semaphore, but that seems less efficient/elegant than just discarding the used semaphore.

Reply:
My question implied that once in the map the entry would never be removed (which I don't think would result in the issue you described). As-is works fine, I think it just a matter of preference. So +1 as-is or leaving it in the map and releasing it in there.

Reply:
That's a fair point, we could leave the semaphore in the map and remove it if an enrich policy is ever deleted. We'll want to ensure that the policy is not being executed currently when performing a policy deletion, so the delete api will most likely have an instance of the locks object anyway. I think the simplicity of leaving it in the map will be worth doing when we make that change